src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompareSymbolsCorrectlyAnalyzer.cs

```diff
@@ -182,22 +184,28 @@ private static void HandleBinaryOperator(in OperationAnalysisContext context, IN
             static bool IsNotInstanceInvocationOrNotOnSymbol(IInvocationOperation invocationOperation, INamedTypeSymbol symbolType)
                 => invocationOperation.Instance is null || IsSymbolType(invocationOperation.Instance, symbolType);
 
-            static bool HasFirstTypeArgumentSymbolType(IMethodSymbol? method, INamedTypeSymbol symbolType)
+            static bool IsBehavingOnSymbolType(IMethodSymbol? method, INamedTypeSymbol symbolType)
             {
                 if (method is null)
                 {
                     return false;
                 }
                 else if (!method.TypeArguments.IsEmpty)
                 {
-                    return IsSymbolType(method.TypeArguments[0], symbolType);
+                    var destinationTypeIndex = method.TypeParameters
+                        .Select((type, index) => type.Name == "TKey" ? index : -1)
+                        .FirstOrDefault(x => x >= 0);
```

Comment:
/!\ This is a bit hacky. /!\ 

I am not sure whether we can safely rely on a generic type parameter name.

As far as I can tell, the non-hacky solution forces us to resolves all overloads, and map them with the index we should look at.

Reply:
I'll let @ryzngard review/approve this change.

Reply:
Ping @ryzngard

Reply:
Looking, sorry for delay

Reply:
Agree this does seem hacky, but maybe no more hacky than depending on ordering. At the end of the day we're at least working on a known set of functions that we don't expect to change drastically, so if it works for the known set maybe it's good enough.

I was looking at the test you wrote without the change, and a surprising piece for me: the type arguments are not in the order I would expect. 

Method display is showing `ILookup<string, IFieldSymbol>` but type arguments are `{ IFieldSymbol, string }`. Going to try and find documentation to see if that's expected.

Reply:
Oh nm, this makes sense. Just got slightly confused with how the information was display. `ToLookup<IFieldSymbol, string>` is the method here so the first type argument is correct. Maybe for things like `ToLookup`, `ToDictionary`, and `GroupBy` we should be looking at the return type and it's typearguments instead of the method signature arguments

Reply:
I can but this will add some complexity to the current solution. Let me know what's your preference.

Reply:
Tagging @ryzngard for last question

Reply:
Ping @ryzngard

Reply:
Ping @ryzngard

Reply:
This looks fine for now